### PR TITLE
Add input validation

### PR DESCRIPTION
**Pull request by Radinyn**

### CODE REVIEW

File: protostar/starknet_gateway/gateway_facade.py [hunk 1]
```diff
@@ -110,20 +117,36 @@ def _load_compiled_contract(self, compiled_contract_path: Path):
...
+                    inputs = from_python_transformer(
+                        abi, fn_name="constructor", mode="inputs"
+                    )(inputs)
+
+                # Validate inputs
+                to_python_transformer(abi, "constructor", "inputs")(inputs)
+            except DataTransformerException as ex:
+                raise InputValidationException(str(ex)) from ex
+
+        return cast(List[int], inputs)
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade.py:149]: Can you add validation in a separate method?

File: tests/integration/gateway/gateway_facade_test.py [hunk 2]
```diff
@@ -101,14 +103,32 @@ async def test_call_to_with_positional_incorrect_args(
...
+    protostar.build_sync()
+    yield protostar.project_root_path / "build" / "main.json"
+
+
+async def test_deploy_fail_input_without_constructor(
+    gateway_facade: GatewayFacade, compiled_contract_without_constructor_path: Path
+):
+    with pytest.raises(InputValidationException) as ex:
+        await gateway_facade.deploy(
+            compiled_contract_without_constructor_path, inputs={"initial_balance": 42}
```
Comment (kasperski95) [inline tests/integration/gateway/gateway_facade_test.py:127]:
```suggestion
            compiled_contract_without_constructor_path, inputs={"UNKNOWN_INPUT": 42}
```

File: tests/integration/gateway/gateway_facade_test.py [hunk 3]
```diff
@@ -117,9 +137,27 @@ def compiled_contract_with_contractor_path_fixture(protostar: ProtostarFixture):
     yield protostar.project_root_path / "build" / "main.json"
 
 
+@pytest.mark.parametrize("inputs", [[42], [{"initial_balance": 42}]])
 async def test_deploy_supports_data_transformer(
-    gateway_facade: GatewayFacade, compiled_contract_with_contractor_path: Path
+    gateway_facade: GatewayFacade,
+    compiled_contract_with_contractor_path: Path,
+    inputs: CairoOrPythonData,
```
Comment (kasperski95) [inline tests/integration/gateway/gateway_facade_test.py:144]: `inputs` are unusued

File: tests/integration/gateway/gateway_facade_test.py [hunk 4]
```diff
@@ -117,9 +137,27 @@ def compiled_contract_with_contractor_path_fixture(protostar: ProtostarFixture):
...
     await gateway_facade.deploy(
         compiled_contract_with_contractor_path, inputs={"initial_balance": 42}
     )
+
+
+async def test_deploy_no_args(
+    gateway_facade: GatewayFacade, compiled_contract_with_contractor_path: Path
+):
+    with pytest.raises(InputValidationException):
+        await gateway_facade.deploy(compiled_contract_with_contractor_path, [])
```
Comment (kasperski95) [inline tests/integration/gateway/gateway_facade_test.py:155]:
```suggestion
        await gateway_facade.deploy(compiled_contract_with_contractor_path)
```

File: protostar/starknet_gateway/gateway_facade.py (hunk 1 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py:149]: The problem is that `from_python_transformer` also performs some validation, so decoupling them is tricky.

File: tests/integration/gateway/gateway_facade_test.py [hunk 5]
```diff
@@ -117,9 +137,27 @@ def compiled_contract_with_contractor_path_fixture(protostar: ProtostarFixture):
...
 
+@pytest.mark.parametrize("inputs", [[42], [{"initial_balance": 42}]])
 async def test_deploy_supports_data_transformer(
-    gateway_facade: GatewayFacade, compiled_contract_with_contractor_path: Path
+    gateway_facade: GatewayFacade,
+    compiled_contract_with_contractor_path: Path,
+    inputs: CairoOrPythonData,
 ):
     await gateway_facade.deploy(
         compiled_contract_with_contractor_path, inputs={"initial_balance": 42}
```
Comment (Radinyn) [inline tests/integration/gateway/gateway_facade_test.py:147]:
```suggestion
        compiled_contract_with_contractor_path, inputs=inputs
```

File: protostar/starknet_gateway/gateway_facade.py [hunk 6]
```diff
@@ -110,20 +115,27 @@ def _load_compiled_contract(self, compiled_contract_path: Path):
...
+        if not inputs:
+            return []
+        if not has_abi_item(abi, "constructor"):
+            if inputs:
+                raise InputValidationException(
+                    "Inputs provided to a contract with no constructor."
+                )
+        else:
+            inputs = transform_and_validate_constructor_inputs_from_python(abi, inputs)
+            validate_cairo_inputs(abi, inputs)
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:136]:
```suggestion
        if not inputs:
            return []
            
        if inputs and not has_abi_item(abi, "constructor"):
            raise InputValidationException(
                "Inputs provided to a contract with no constructor."
            )

        inputs = transform_and_validate_constructor_inputs_from_python(abi, inputs)
        validate_cairo_inputs(abi, inputs)
```

this one is much more understandable at quick glance to me

**Review by mkaput — APPROVED** [review]

File: protostar/starknet_gateway/gateway_facade.py (hunk 1 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py]: done

File: protostar/starknet_gateway/gateway_facade.py (hunk 6 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py]: We need to prevent checking inputs to a constructor if it doesn't exist to avoid a LookupError.

Comment (mkaput) [reply protostar/starknet_gateway/gateway_facade.py]: The only thing I did in this diff is restructuring code, I haven't changed the logic at all :)

File: protostar/starknet_gateway/gateway_facade.py [hunk 7]
```diff
@@ -393,3 +420,26 @@ def __init__(self, compilation_output_filepath: Path):
             "Did you run `protostar build`?"
         )
         self._compilation_output_filepath = compilation_output_filepath
+
+
+def transform_and_validate_constructor_inputs_from_python(
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade.py:425]:
```suggestion
def transform_constructor_inputs_from_python(
```

File: protostar/starknet_gateway/gateway_facade.py (hunk 6 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py]: Sorry, `return []` is a bug that appeared during the refactor, it should be `inputs = []` (to also validate constructor with no inputs), when that's the case the logic is different.

File: protostar/starknet_gateway/gateway_facade.py [hunk 8]
```diff
@@ -110,20 +115,27 @@ def _load_compiled_contract(self, compiled_contract_path: Path):
...
+        if not has_abi_item(abi, "constructor"):
+            if inputs:
+                raise InputValidationException(
+                    "Inputs provided to a contract with no constructor."
+                )
+        else:
+            inputs = transform_and_validate_constructor_inputs_from_python(abi, inputs)
+            validate_cairo_inputs(abi, inputs)
+
+        return cast(List[int], inputs)
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade.py:138]: Code should be written in a such way that casting is not necessary.

File: protostar/starknet_gateway/gateway_facade.py (hunk 6 above)
Comment (mkaput) [reply protostar/starknet_gateway/gateway_facade.py]: updated the snippet

**Review by kasperski95 — APPROVED** [review]
Not fan of this code, but it's internal to `gateway_facade` so approving.

**Pull request merged**